infrastructure/data_access/http/synthetics_repo.py

```diff
@@ -56,9 +62,35 @@ def _get_mesh_test_rows(self, test_id: TestID, results_lookback_seconds: int) ->
         return transform_to_internal_mesh_rows(most_recent_result)
 
     def _get_agents(self, test_id) -> Agents:
+        agent_ids = self._get_agent_ids(test_id)
+
+        with self._cached_agents_lock:
+            self._update_agents_cache_if_needed(agent_ids)
+            return self._cached_agents.filter(agent_ids)
+
+    def _get_agent_ids(self, test_id):
+        logger.debug(f"Getting agent IDs for test with ID {test_id}")
         test_resp = self._api_client.synthetics_admin_service.test_get(test_id)
+        agent_ids = test_resp.test.settings.agent_ids
+        logger.debug(f"Got agent IDs for test with ID {test_id}: {agent_ids}")
+        return agent_ids
+
+    def _update_agents_cache_if_needed(self, agent_ids):
+        if self._is_cached_agents_update_necessary(agent_ids):
+            self._update_agents_cache()
+
+    def _is_cached_agents_update_necessary(self, agent_ids: List[AgentID]) -> bool:
+        cached_agent_ids = self._cached_agents.list_agent_ids()
+        for agent_id in agent_ids:
+            if agent_id not in cached_agent_ids:
+                return True
+        return False
+
+    def _update_agents_cache(self) -> None:
+        logger.debug("Updating agents cache")
         agents_resp = self._api_client.synthetics_admin_service.agents_list()
-        return make_internal_agents(agents_resp.agents, test_resp.test.settings.agent_ids)
+        self._cached_agents = transform_to_internal_agents(agents_resp.agents)
```

Comment:
Hold the lock around this statement. Assignments are not guaranteed to be atomic in Python (basically because every variable is an entry in the namespace dictionary and dictionary operations are not atomic). See https://github.com/google/styleguide/blob/91d6e367e384b0d8aaaf7ce95029514fcdf38651/pyguide.md#218-threading

Reply:
It's important to keep in mind that the methods stored in main.py "self._routes" can be called from multiple threads. Maybe it's worth to highlight this in the code, what do you think?

Reply:
Yes, put a comment to that effect in the code.